### ISSUE TITLE
Some notes about cyclic dependencies

### ISSUE THREAD

**Issue by brupelo** [kayhayen/Nuitka#88]
Today I've been trying to grasp the high-level view of the project so I've reviewed all modules. I have noticed the project graph is not a DAG(directed acyclic graph) but a DCG(directed cyclic graph). In order to make this analysis I haven't considered any external code to be dependencies, that is, code that's tested by other people (python modules, pyqt5, ...) so we can assume is bug-free. A first pass would be something like:

    codegen
        nuitka.__past__
        nuitka.Builtins
        nuitka.Constants
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.PythonVersions
        nuitka.Tracing
        nuitka.utils
        nuitka.Version

    containers
        nuitka.PythonVersions

    distutils
        nuitka.importing
        nuitka.Version

    finalizations
        nuitka.Options
        nuitka.Tracing
        nuitka.__past__
        nuitka.importing
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.tree

    freezer
        nuitka.__past__
        nuitka.codegen
        nuitka.containers
        nuitka.importing
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.tree
        nuitka.utils

    gui
        nuitka.__past__
        nuitka.codegen
        nuitka.containers
        nuitka.importing
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.tree
        nuitka.utils

    importing
        nuitka.__past__
        nuitka.codegen
        nuitka.containers
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.tree
        nuitka.utils

    nodes
        nuitka.__past__
        nuitka.Builtins
        nuitka.codegen
        nuitka.Constants
        nuitka.containers
        nuitka.importing
        nuitka.ModuleRegistry
        nuitka.optimizations
        nuitka.Options
        nuitka.PythonOperators
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.tree
        nuitka.TreeXML
        nuitka.utils
        nuitka.Variables

    optimizations
        nuitka.__past__
        nuitka.Builtins
        nuitka.containers
        nuitka.importing
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.Tracing
        nuitka.tree
        nuitka.TreeXML
        nuitka.utils
        nuitka.Variables

    plugins
        nuitka.__past__
        nuitka.Builtins
        nuitka.containers
        nuitka.importing
        nuitka.ModuleRegistry
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.tree
        nuitka.TreeXML
        nuitka.utils
        nuitka.Variables

    tools <- Not analized

    tree
        nuitka.__past__
        nuitka.Constants
        nuitka.importing
        nuitka.nodes
        nuitka.Options
        nuitka.plugins
        nuitka.PythonVersions
        nuitka.SourceCodeReferences
        nuitka.Tracing
        nuitka.utils

    utils
        nuitka.Options
        nuitka.PythonVersions
        nuitka.Tracing

If we try to simplify the above and we consider files living in root (ie: `__past__.py, Builtins.py, Constants.py, MainControl, ..., Version.py` ) we'd get a DCG like:

    codegen -> {nodes, root, utils}
    containers -> {root}
    distutils -> {importing, root}
    finalizations -> {importing, plugins, root, tree}
    freezer -> {codegen, containers, importing, nodes, plugins, root, tree, utils}
    gui -> {root, codegen, containers, importing, nodes, plugins, tree, utils}
    importing -> {codegen, containers, nodes, plugins, root, tree, utils}
    nodes -> {root, codegen, containers, importing, optimizations, tree, utils}
    optimizations -> {root, containers, importing, nodes, plugins, tree, utils}
    plugins -> {root, containers, importing, nodes, tree, utils}
    tree -> {root, importing, nodes, plugins, utils}
    utils {root}

I say DCG cos if you look at it you'll notice cycles such as:

    cycles = [
        (codegen, nodes),
        (importing, nodes),
        (importing, plugins),
        (importing, tree),
        (nodes, optimizations),
        (nodes, tree),
        (plugins, tree)
    ]

Not proposing anything about this one but I've thought it was worth posting my review somewhere as I'm deleting the notes :D. Actually the was creating the graph with some node editor but it was a ton of work drawing the whole thing so I gave up ;) . 

Related to this topic, did you read [Lakos' book](https://www.amazon.com/Large-Scale-Software-Design-John-Lakos/dp/0201633620)? It gives a lot of advices from the point of view of considering large projects like graphs, so refactoring when you think that way becomes really easy and automatized. It's a really nice reading (not just for c++ but in general) :)

Yours,
B

**Comment by brupelo** [comment]
Related stuff [http://structure101.com/blog/2014/03/why-cyclic-dependencies-explode-complexity/](http://structure101.com/blog/2014/03/why-cyclic-dependencies-explode-complexity/)

**Comment by kayhayen** [comment]
Hello Bruno,

getting a critical review like that really made my day. Very exciting. I
often wished I would get more of that.

White replying I got into a fair bit of output and explaining yesterday
night. I will break it down into multiple replies of actual quality, i.e.
explaining things in ways actually understandable, coming over the next day
or two. That way it won't be a monster thread, and something I can turn
into actual blog posts in the end.

Yours,
Kay

**Comment by kayhayen** [comment]
So, first of all, I recalled that I once did that kind of analysis, but with actual graphs. There is a snakefood package on pip, that I used, and there is misc/generate-dependency-graph.py or so, that is making a graph out of things, but it needs severe updating. I am pushing something for factory later today.

Then, plugins are suppose to hook everything, and will introduce technically cyclic dependencies no matter what. They create their own modules on the fly, change how importing works, while using importing to locate stuff on their own. There is no way that wouldn't be a cyclic dependency. So we need to ignore those. It must be that ugly, but since Plugins are supposed to be no-ops normally, I think they do not place that much of a burden, and they are the way to shoot into your feet.

I am following up some more later. I was ill unfortunately and have some catching up to do.
